modules/java/native/java_machine.c

```diff
@@ -113,6 +113,12 @@ void
 java_machine_unref(JavaVMSingleton *self)
 {
   g_assert(self == global_jvm);
+  /* The last reference is always hold by a AH_SHUTDOWN app hook, when there is no java configured.
+   * The only way to get ride of it to shutdown syslog-ng.  */
+  if (g_atomic_counter_racy_get(&self->ref_cnt) == 1)
```

Comment:
I wouldn't access an atomic counter in a racy way even if atomicity is guaranteed here.
We had difficulties to trace such things after the code changed a few years later.

I would also remove `g_atomic_counter_racy_get` completely.